Improve Map Type Stability

willtebbutt opened the pull request:
As the title suggests, improves the type-stability of map. Does so by getting rid of a branch. Tests pass locally, and I've added a few more.

DhairyaLGandhi reviewed (approved):
Lgtm, can we add another test case for non empty and empty collections too?

willtebbutt commented:
> Lgtm, can we add another test case for non empty and empty collections too?

Sorry, I'm not sure what kind of thing you have in mind here -- could you elaborate?

willtebbutt commented:
Hmmm confusingly, I can't replicate the error I'm seeing locally...

willtebbutt commented:
> Sorry, I'm not sure what kind of thing you have in mind here -- could you elaborate?

Added additional tests for type-stability.

> Hmmm confusingly, I can't replicate the error I'm seeing locally...

Have fixed this and added unit tests designed to catch the specific edge case on which map was failing.

Will bors when I know that CI passes.

willtebbutt commented:
bors r+
